### PR TITLE
[SiVal] I2C test plan

### PR DESCRIPTION
**Pull request by marnovandermaas**
This is an initial version of the I2C silicon validation test plan. It takes inspiration from the counterpart in key manager: https://github.com/lowRISC/opentitan/pull/19521

I tried to create enough tests so that all of I2C's features are covered, but it would probably be good to get the input of more people here. Alternatively, we can merge this as a base version and iterate over it with future PRs.

### CODE REVIEW

**Comment by marnovandermaas** [comment]
Note: I will be away next week, so if any changes need to be handled in that time someone else will need to pick it up.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson
Comment (moidx) [inline hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:71]: nit: `stage: V3` to avoid lowering the current regression completion results. Same for `chip_sw_i2c_override`.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson
Comment (moidx) [inline hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:7]: Should we add an explicit check for clock stretching to one of the tests?

**Review by moidx — APPROVED** [review]
Thanks @marnovandermaas! I added a couple of small comments. I think it is ok to submit this PR and then work on follow ups if needed. Thanks!

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson
Comment (engdoreis) [inline hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:94]: New line missing.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson
Comment (engdoreis) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:7]: Clock stretching is important, but I also suggest adding a test for NAK detection.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson
Comment (engdoreis) [inline hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:61]: I suggest a loopback test where the host writes X and reads X+1 or X>>1.
Should we describe here how the baud rate will be measured (Since i2c is synchronous, the fact the data matches doesn't mean the speed is correct)?

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson
Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson]: Thanks for noticing this! This should now be fixed.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson
Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:7]: Thanks for these suggestions, I tried to avoid specifying the whole I2C protocol in the feature list, but I agree that clock stretching and nack detection are important corner cases to test. So I have added them to the feature list and added two tests for them. Let me know what you think of them.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson
Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:61]: Thanks for this suggestion. I've made an initial attempt at this, but happy to adjust it based on further feedback.

**Comment by marnovandermaas** [comment]
CHANGE AUTHORIZED: hw/ip/i2c/data/i2c.hjson

**Comment by marnovandermaas** [comment]
Since this PR is quite time sensitive, I propose that we merge this PR once CI passes and make any changes in follow-up PRs. Can another committer please authorize these changes:

CHANGE AUTHORIZED: hw/ip/i2c/data/i2c.hjson

File: hw/ip/i2c/data/i2c.hjson
Comment (crteja) [inline hw/ip/i2c/data/i2c.hjson:163]: It would be good to add RSTART also as we have observed an issue with it.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson
Comment (crteja) [inline hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:110]: [Clock stretching](https://opentitan.org/book/hw/ip/i2c/doc/theory_of_operation.html#clock-stretching) is supported in target mode as well and it would be good to have a separate test case for it.

Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:110]: I specified this host test because during SiVal testing the testbench has control over the target and can do the clockstretching. How do you force the OpenTitan target to do clock stretching, even when the test bench only has control over the host?

File: hw/ip/i2c/data/i2c.hjson
Comment (marnovandermaas) [reply hw/ip/i2c/data/i2c.hjson:163]: Thanks added repeated start as a feature and a test for it.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson
Comment (crteja) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:110]: Agreed but I2C supports clock stretching in target mode as well (and is allowed in I2C protocol). For instance, when ACQ FIFO is full and Ibex is busy with other higher priority interrupts and delays reading from the FIFO, OT I2C target holds SCL low until ACQ FIFO has space for additional bytes. This scenario can be triggered in SiVal.

`chip_sw_i2c_clockstretching` only focuses on OT I2C host mode. As each I2C IP instance in OT is used either as a Host or a Target, , there should be a separate target mode test that covers clock stretching scenarios mentioned in https://opentitan.org/book/hw/ip/i2c/doc/theory_of_operation.html#clock-stretching. 
There is a DV test that covers some of these scenarios https://opentitan.org/book/hw/ip/i2c/data/i2c_testplan.html#target_clock_stretch

Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:117]: Thank Ravi! I've made an initial attempt at this. Happy to change it though.

Comment (crteja) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:117]: Thanks Marno. This looks good for basic clock stretch in target mode. Do we need to exhaustively cover all cases of clock stretch in SiVal ?

Comment (johngt) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:117]: @crteja / @marnovandermaas 
For SiVal Phase-II we should cover base cases.
For SiVal Phase-III this is where we should cover exhaustive testing

Comment (crteja) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:117]: @marnovandermaas , would it make sense to have a separate SV3 test that covers all cases of clock stretching and mark this test as SV2 ?

Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:117]: Ok, I've marked this as SV2 again.

**Review by rswarbrick — COMMENTED** [review]
CHANGE AUTHORIZED: hw/ip/i2c/data/i2c.hjson

**Comment by rswarbrick** [comment]
CHANGE AUTHORIZED: hw/ip/i2c/data/i2c.hjson

**Pull request merged**
